src/components/QueryEditor.tsx

```diff
@@ -71,6 +71,42 @@ const QueryClause = ({
     () => sourceToTargetPlaceholder({ relation: con.relation }),
     [con.source, con.relation]
   );
+  const setConditionRelation = useCallback(
+    (e: string) => {
+      window.clearTimeout(debounceRef.current);
+      setConditions((_conditions) =>
+        _conditions.map((c) =>
+          c.uid === con.uid ? { ...con, relation: e } : c
+        )
+      );
+      debounceRef.current = window.setTimeout(() => {
+        setInputSetting({
+          blockUid: con.uid,
+          key: "Relation",
+          value: e,
+          index: 1,
+        });
+      }, 1000);
+    },
+    [setConditions]
+  );
+  const setConditionTarget = useCallback(
+    (e) => {
+      window.clearTimeout(debounceRef.current);
+      setConditions((_conditions) =>
+        _conditions.map((c) => (c.uid === con.uid ? { ...con, target: e } : c))
+      );
```

Comment:
One cool thing I discovered by stalking tldraw's code is that React's useState setters can accept a function, where the old value is an input. This is huge for memoization, since you no longer need to create a new function for every new state value